pkg/node/secret.go

```
@@ -298,82 +297,35 @@ func (n *Node) deleteSecret(ctx echo.Context) error {
+		// Check each physical node's heartbeat
+		if n.checkHeartbeat(key) {
+
+			// Send each physical node a message asking for all secrets within that role's scope
+			request := &message.Request{
+				From: n.Pid,
+				To:   key,
+				Code: message.DELETE_SECRET_ALL_INSTANCES,
+				Payload: map[string]interface{}{
+					"key": secretHash,
+				},
+			}
 
-	var reply message.Reply
-	err = message.SendMessage(targetNodeAddr, "Node.DeleteSecret", request, &reply)
-	if err != nil {
-		// When owner node is down, we still need to try delete all replicas
-		relayVirtualNodes, err := n.getRelayVirtualNodes(targetLocation)
-		if err != nil {
-			return ctx.JSON(http.StatusInternalServerError, &api.Response{
-				Success: false,
-				Error:   err.Error(),
-				Data:    nil,
-			})
-		}
-		config, err := config.GetConfig()
-		if err != nil {
-			return ctx.JSON(http.StatusInternalServerError, &api.Response{
-				Success: false,
-				Error:   err.Error(),
-				Data:    nil,
-			})
-		}
-		err = n.relaySecretDeletion(config.ConfigNode.ReplicationFactor, strconv.Itoa(secretHash), relayVirtualNodes)
-		if err != nil {
-			return ctx.JSON(http.StatusInternalServerError, &api.Response{
-				Success: false,
-				Error:   err.Error(),
-				Data:    nil,
-			})
-		}
-		return ctx.JSON(http.StatusOK, &api.Response{
-			Success: true,
-		})
-	} else {
-		replyPayload := reply.Payload.(map[string]interface{})
-		success := replyPayload["success"].(bool)
-		if success {
-			return ctx.JSON(http.StatusOK, &api.Response{
-				Success: true,
-			})
-		} else {
-			err := replyPayload["error"].(error)
-			return ctx.JSON(http.StatusInternalServerError, &api.Response{
-				Success: false,
-				Error:   err.Error(),
-				Data:    nil,
-			})
+			var reply message.Reply
+			err := message.SendMessage(value, "Node.DeleteSecret", request, &reply)
+			if err != nil {
+				return ctx.JSON(http.StatusInternalServerError, &api.Response{
+					Success: false,
+					Error:   "One of the node is down ",
```

Same as #32 , if a node is down, we assume the data is wiped out clean, hence, we can still guarantee the client that all secrets are deleted instead of stopping here and returning an error to the client. So I think here don't need to return an error, but continue deleting the rest of the secrets, then can just keep track of whatever nodes are down and return this information to the client in the `Data` field in the response.

---

Alright, implemented the logic that I used for the get all secrets flow.